Class dependencies

gkossakowski opened the pull request:
This PR implements class-based dependency tracking. The basic idea is simple: let's track dependencies at class level instead of at source level. Dependencies on inner classes are tracked separately. Also, APIs (to detect which dependencies has changed) are tracked for each class separately.

The detailed list of what has been done and what's left to be done was tracked in sbt/sbt#1104. I'm pasting it here for convenience.

----

### Tasks

Class-based dependency tracking can be implemented in two phases. The third phase is dedicated to testing on large projects.
#### Invalidate by inheritance based on class dependencies (phase 1)

_~~Tasks marked as completed are completed only in a prototype implementation. No changes has been merged to sbt yet.~~_
- [x] add tracking of declared classes in source files; we want to invalidate classes but at the end we need className -> srcFile relation because we can recompile source files only
- [x] add dependency tracking of top-level (not owned by any other class) imports; we'll assign dependencies introduced by those imports to the first top level class/object/trait declared in the source file (see notes below for details)
- [x] track dependencies at class level (and use `declaredClasses` to immediately map them back to files until the rest of the algorithm is implemented)
- [x] add tracking of children of sealed parents; we should track this in `API` object corresponding to sealed parent; this will enable us to perform proper invalidation when a new case (class) introduced to a sealed hierarchy
- [x] add per (top level) class tracking of api hashes; tracking of api hashes per each (even nested) class separately will be done in phase 2
- [x] handle dependencies coming from local and anonymous classes (see discussion below)
- [x] switch invalidation of dependencies by inheritance to invalidate at class level instead of at source level (would fix #2320)
- [x] distinguish between source and binary class names. Introduce a new relation `binaryClassName: Relation[String, String]` and use it whenever the conversion is needed.
#### Track name hashes per class (phase 2)
- [x] refactor tracking of APIs to be done at class level instead of at source level
- [x] extract APIs of all classes (including inner ones) separately
- [x] track name hashes at class level instead of being at source level (see also: #2319)
- [x] implement the rest of invalidation logic (member reference, name hashing) based on class names (get rid of mapping through `declaredClasses` in almost entire algorithm) (would fix #2320)

Only once the last bullet is merged we'll see improvement to incremental compilation when big number of nested classes is involved (e.g. like in Scala compiler's case).
#### Testing, bug fixing and benchmarking (phase 3)
- [x] test with the Scalatest repo
- [x] test with the scala repo using the sbt build
- [x] test with the specs2 repo
- [ ] fix handling of anonymous and local classes defined in Java (similar logic will have to be implemented as for handling local Scala classes)
- [ ] index classes only declared (not inherited) in source files
- [ ] benchmark `;clean;compile` performance
- [x] simplify classpath and Analysis instance lookup in incremental compiler (I think number of classpath lookups can be reduced now)
#### Merge changes upstrem, prepare for shipping (phase 4)
- [x] determine the location where this work will be merged into
### Benefits

The improved dependency tracking delivers up to 40x speedups of incremental compilation in scenarios tested. Check benchmarking results here: https://github.com/sbt/sbt/issues/1104#issuecomment-193002190

The speedups are caused by fixing two main issues:
1. The #2319 would be fixed once name hashes are tracked per class. This way introduction of a new class and members coming with it would not affect source files dependent (by member ref) on already existing classes.
2. Effects of adding members (like methods) to a class would affect only classes that inherit from that class. At the moment, adding a member to a class that nobody inherits from can trigger invalidation of all descendants of all classes defined in the same source file (see #2320 for a specific scenario).

The 1. is likely to be triggered by any code base that uses more than one class defined in a single source file. The 2. is affecting code bases with big number of nested classes that are inherited.
One example is Scala compiler itself. Even with name hashing, we invalidate too much and code edit cycle becomes long whenever a new member is introduced.

----

### Known issues
- Java-specific tests are ignored (see #83)
- Java Anonymous and local classes are not handled properly leading to either crashes or undercompilation (see the task above)
- asSeenFrom is not honored correctly for inherited classes (they are extracted only as seen from declaration site) which leads to under compilation and missing typechecking errors (https://gitter.im/scala/contributors?at=56ec053bbb4a1731739afb6f)
- handling of private members in traits has been reverted due to sbt/sbt#2490
#### Backwards compatibility

This PR affects backwards compatibility in several ways:
- support for Scala 2.7, 2.8 and 2.9 has been dropped
- support for `Analysis.groupBy` operation has been dropped (it's not clear who needs it)
- the API of `Analysis` object is very different now because most methods take classNames as arguments instead of file names

gkossakowski committed:
Add an explicit type annotions to the def `allRelations`.

The inferred type is not great so let's enforce the right existential
explicitly.

gkossakowski committed:
Remove MRelationsDefaultImpl

Remove implementation of Relations that supports the old incremental
compilation algorithm. This will simplify changing relations to be
class-based.

gkossakowski committed:
Remove deprecated methods from Relations.

gkossakowski committed:
Remove deprecated methods from Analysis.

gkossakowski committed:
Remove dead code from AnalysisFormat.

Since introduction of TextAnalysisFormat, most of AnalysisFormat is
dead code. Some bits seem to be still used in various parts of sbt, though.

Updated TextAnalysisFormatSpecification to not use deprecated methods.

gkossakowski committed:
Record declared class names in source files.

Introduce a new relation declaredClasses in Relations. It tracks names
of classes declared in given source file. Names of classes are fully
qualified names as seen at pickler phase (i.e. before flattening).
Objects are represented as object name with dollar sign appended to it.

The `declaredClasses` relation will be needed to map invalidated classes
back to source files where they are defined. It's worth mentioning that
there's a relation called `classes` defined already. However, that relation
tracks names of classes as seen by the Scala compiler backend. These are
names that will later appear in bytecode. For our purposes, we want to
have names of classes as they are declared in source code.

Declared classes are tracked properly for both Scala and Java files.

While updating ScalaCompilerForUnitTesting I flipped the nameHashing flag
to be true by default. That's in sync with sbt's default value for that
flag now.

gkossakowski committed:
Log progress on recording class-based dependencies.

Everything compiles but tests are failing. This commit should be split into two parts:
  - refactoring of Relations, TextAnalysisFormat, etc. that introduces support for relations where first element in their
     pair is not a file. At the moment, there're many places where code assumes that all relations are Relation[File, T].
   - the rest of the code

gkossakowski committed:
Remove AnalysisTest.

Remove AnalysisTest that tests functionality not supported by name hashing.

gkossakowski committed:
Dependency tracking in locally defined classes requires more work.

Dependencies coming from classes defined locally are not tracked properly
and proper accounting of such dependencies will require refactoring
of the Dependency phase. We'll need to make an explicit decision where
dependencies of locally defined classes should go. They cannot be tracked
at the level of a class that introduces the dependency because we track
only classes publicly visible from other compilation units (this is
left to be defined precisely later on).

As for now, we just mark a test in DependencySpecification as pending.

gkossakowski committed:
Do not crash when source doesn't declare any classes.

Just get an empty set instead.

gkossakowski committed:
Fix bugs in className->src resolution.

The helper method in IncrementalNameHashing was using a wrong
relation and had bugs in logic emitting fatal error messages.

gkossakowski committed:
Fix treatment of refinements in dependency extraction.

Refinements are represented as classes internally but we want to
record dependencies on named classes only. Therefore, we ignore
the refinement class and only look at symbols referred from the
refinement.

gkossakowski committed:
Improve class->src mapping of dependencies.

Introduce a dedicated ClassToSourceMapper and use it throughout the
incremental compiler. The ClassToSourceMapper takes care of properly
mapping classes that have been moved between source files by looking
at relations before recompilation and after recompilation.

gkossakowski committed:
Remove use of scala.Symbol from ScalaCompilerForUnitTesting

We used to use scala.Symbol to easily identify sources in file system
independent manner. Since we switched to class-based dependency tracking,
we can use class names directly.

gkossakowski committed:
Extract dependencies on objects properly.

Object's members are declared in module class but references to objects
in trees are typechecked with module symbol assigned to it. We handle
references to objects by mapping module symbols to module class symbols.

gkossakowski committed:
Fix bug in mapping source dep into class dep.

We would forget to consider declared classes for mapping. This would cause
invalid mapping in case of mixed compilation (of java and scala files).

gkossakowski committed:
Fix invalidation by external changes.

This commit fixes invalidation caused by api changes to an external
(defined in other project) class. The problem was that we tried to
convert internal class-> external class dependency relation to
src->src but we would end up with an empty relation because we can't map
external class to a src (it's defined outside of current project).

We fix this problem by converting relation to internal src -> external
class. This conversion is always possible.

We also fix a bug in ClassToSourceMapper where conversion would lose some
relations in case the left side of relation (class name) mapped to
several different source files. This is possible when class is moved
between source files.

gkossakowski committed:
Remove old incremental compilation algorithm implementation.

We do not support it in a class-based dependency tracking.
Remove tests that are specific to the old incremental compiler and are
failing now.

gkossakowski committed:
Switch from `declaredClasses` to `classes` relation.

The `classes` relation does the same thing as `declaredClasses` relation
except it uses different string representation for class names. The
`classes` uses names as seen after flatten phase but `declaredClasses`
looks at symbols at pickler phase.

I think the choice doesn't matter as long as the same naming scheme is
used consisstently everywhere. Let's give flattened class names a try
and find out if we can remove `declaredClasses`.

gkossakowski committed:
Hash individual top level definitions

Modify xsbti.api.Source to carry api hash separately for each
top level definition. Modify api extraction phase to compute hashes for
each top-level definition separately.

This will enable us to invalidate individual top-level classes.

gkossakowski committed:
Invalidate dependencies at class level.

This commit implements invalidation of dependencies at class level both
for inheritance and memberRef dependencies. However, information that
determines api changes is still tracked mostly at source file level.

Name hashes are tracked at source level (classes considered for
invalidation are mapped back to source files before name hash lookup).
For that reason, #2319 is still not fixed.

Api hashes are tracked per-class but only for top-level classes. It's not
a final design but a quick hack to prove that class-based invalidation is
working correctly. It was enough to scenario described in #2320 fixed.
This is shown in `class-based-inheritance` scripted test.
The final design of api hashing will be that all publicly accessible
classes will get their own separate hash.

The sealed test is marked pending as described in notes of #1104.

Other thing to note is that we're getting close to being able to strip
abstractions that deal with the fact that dependencies were tracked either
as Relation[File, File] (for internal dependencies) or
Relation[File, String] (for external dependencies). We track all
dependencies uniformly as Relation[String, String] now.

gkossakowski committed:
Use flatnames for classfile lookup.

We tried to switch to src class names everywhere in the incremental
compiler but that broke the logic that was looking up generated class
files. As a result, we wouldn't record inner classes properly due to
difference between src and flat class names.

I've added a scripted test that verifies proper recording of inner classes.

gkossakowski committed:
Distinguish between binary and source class names.

This is a large commit so I'll describe high level changes first and then
go through each individual source file.

At the center of this commit is differentation between binary and source
class name. The incremental compiler needs to track both. The primary way
of tracking classes and referring to them is through source (canonical
in Java's terminology) class names. Binary names are needed for classpath
lookups and turning binary dependencies into source dependencies. We
perform that process each time there's a dependency on a class that has
been compiled in different compiler run and from current compiler run is
seen as a binary dependency (dependency on a class file).

We maintain mapping between source and binary class names in
`Relations.classes` relation. We track in that relation only names of
non-local classes. Non-local casses are the ones that can be referred
from other source files. The top-level, public inner, package private,
protected are all examples of non-local classes. Locals classes are ones
defined within a scope not owned by a class (e.g. in a block or a method)
or anonymous classes. We maintain names of non-local classes because only
those classes have canonical names and only those can be dependencies of
other classes (hence we might need to perform mapping between binary and
source name).

Most changes in this commit are driven by changes to AnalysisCallback,
the Java interface. I changed indentation from tabs to spaces in that file
so the diff is hard to read. Let me explain the most important changes:

  - classDependency and binaryDependency methods have clear names of
    parameters that indicate whether they expects binary or source class
    names
  - generatedClass has been split into two: generatedNonLocalClass and
    generatedLocalClass. The first one takes both forms of the name as
    parameters, whereas the second variant takes just source file and
    class file and records just that. We used to pass some garbage names
    for local classes before. Now we make a clear distinction in
    AnalysisCallback between local and non-local classes and amount of
    information that is tracked about both. This change reduces the size
    of Analysis object and its serialized form. The generatedNonLocalClass
    method overlaps with declaredClass method. I haven't decided what to
    do about it yet but I think declaredClasses can be merged into
    generatedNonLocalClass

Change to `generatedClass` caused a little bit different tracking of
compilation products (class files). We do not track a binary name for
every class file produced; we track it only for class files corresponding
to non-local classes. For that reason we split products into products
and classes in Analysis and Relations.

All changes to Compile.scala (AnalysisCallback implementation) are about
changes to how names and classes are tacked. They are all stragithforward
changes.

Analyzer makes it more clear what kind of names it's tracking and passing
to AnalysisCallback.

Analyze, the class that extracts dependencies from class files compiled
by Java compiler has been adapted to changes in AnalysisCallback and
clarifies when binary and source class names are used. I added a test,
AnalyzeSpecification, that checks whether class dependencies on and from
inner classes are extracted properly.

I also added a scripted test that tests the same thing but verifies
information recorded in Analysis. The reason for a duplicated test is that
scripted test checks a whole lot more:

  - it verifies a correct behavior of AnalysisCallback implementation
  - it verifies that adding class dependencies is routed properly in
    Analysis data structures

Scripted test is an integration test whereas AnalyzeSpecification is a unit
test. In order to implement AnalyzeSpecification I had to introduce
a class that lets me compile Java source on a fly:
JavaCompilerForUnitTesting.

The common functionality between java and scala compiler for unit testing
that extracts data from TestCallback has been pushed to its companion
object. For that to work, I had to add dependency on interfaceProj's test
scope from classfileProj in build.sbt.

gkossakowski committed:
Clarify source and binary class name handling.

Introduce ClassName trait that centralizes logic for creating both source
and binary names for a class symbol. In addition to using it in Analyzer,
Dependency, ExtractDeclaredClasses we also use it in ExtractAPI. We want
every component of the incremental compiler to use consitent naming scheme.

Add ClassNameSpecification that documents expected behavior of ClassName.

gkossakowski committed:
Introduce `binaryClassName` relation.

The `binaryClassName` relation maintains mapping between source and binary
class names. This mapping is needed to map binary dependencies back to
source dependencies in case of separate compilation (where we see
dependencies on class files). You can see that mapping being used in
`binaryDependency` method implementation of Analysis callback. Previously,
we would map class file to a source file it was produced from and then
assume that dependency is on any (all) of classes declared in that class.
Introduction of `binaryClassName` lets us map dependency back to source
class name directly and remove that imprecision of dependency tracking.

We maintain mapping between source and binary class names just for
non-local classes. Check this
https://github.com/sbt/sbt/issues/1104#issuecomment-169146039 for the
discussion of local and non-local classes.

We also rework tracking of products in Analysis by introducing explicitly
the concept of local and non-local products corresponding to local and
non-local classes. This helps us to clarify for which classes we track
source and binary class names.

gkossakowski committed:
Add handling of unused top level imports

Implements a strategy for recording dependencies introduced by top level
imports by assigning those dependencies to the first top level class.
In case there are top level imports but no top level class/trait/object
defined in a compilation unit, a warning is issued. The rationale for this
strategy can be found at:
https://github.com/sbt/sbt/issues/1104#issuecomment-174195925

Add an unit test covering different cases of top level imports (e.g.
defined in nested packages).

Mark the scripted test source-dependencies/import-class as passing after
a small modification of adding a top level class.

gkossakowski committed:
Change indentation of DependencyContext.java to spaces.

gkossakowski committed:
Handle dependencies coming from local classes.

Dependencies introduced by local classes require special handling because
we track only non local classes (that can be referred from other files)
in dependency relations. To overcome this problem, dependencies introduced
by a local class are recorded as introduced by an outer class that is
non local. However, this introduces a new problem with dependencies
introduced by inheritance. We don't want local inheritance dependencies
to cause a transitive invalidation of all classes that inherit from the
outer class containing the local class. Check the comment in
Relations.scala this patches introduces or follow the discussion of this
problem at: https://github.com/sbt/sbt/issues/1104#issuecomment-169146039

To capture the subtlety of inheritance dependencies from local classes,
we introduce `LocalDependencyByInheritance` case to `DependencyContext`
enum.

TestCallback has been modified to return extracted local inheritance
dependencies and a test in DependencySpecification has been updated
accordingly.

The Dependency phase has been reworked to handle local classes properly
by mapping dependencies to outer, non local classes.Check the
implementation for details of the mapping. It's worth mentioning that
mapping is implemented as an amortized O(1) lookup so this change
doesn't affect performance of extraction phase negatively.

The invalidation logic has been modified to take into account inheritance
dependencies introduced by local classes. The patch is small because
the invalidation logic is very straightforward: we invalidate local
inheritance dependencies non-transitively and we do not apply name hashing
dependency pruning.

Lastly, we mark local-class-inheritance scripted test as passing.

gkossakowski committed:
Fix formatting of a comment in NameHashingSpecification

Otherwise, scalariform reformats this file on every `compile`
invocation.

gkossakowski committed:
Cleanup imports in Dependency.scala

gkossakowski committed:
Fix invalidation of sealed class hierarchies

With introduction of class-based dependency tracking, sealed class
hierarchies need special attention. Introduction of a new class that
inherits from a sealed class affects exhaustivity checking of patterns
involving parent class. In other words, a newly introduced class can
affect treatment of an existing class.

Fortunately enough, by definition of a sealed modifier, the list of all
children of a sealed class is available. We need to store that information
and invalidate the parent class when the list of its children changes.
We store the collection of children as part of parent's API representation.

Therefore, the `ClassLike` is modified to hold the collection
`childrenOfSealedClass` that is always empty if a class is not a sealed
class. Elements of that collection are kept in sorted order to ensure
stability. Classes that operate on API objects are updated:
APIUtil, HashAPI and SameAPI.

There's a new test in ExtractAPISpecification that checks if changes to
sealed hierarchy are detected as changes to API object corresponding to
a parent class.

The ClassToAPI that creates API representation of Java class files using
reflection is updated to handle Java enums as sealed hiearchies.
This mimics Scala compiler's treatment of Java enums.

Lastly, the scripted `sealed` test is marked as passing.

gkossakowski committed:
Add pending tests for recording local classes

Local and anonymous classes should not be recorded but they are at the
moment. This commit adds pending tests that check if local and anonymous
classes are recorded.

Both unit tests and scripted tests are added. The unit test checks if
Analyzer compiler phase works correctly. The scripted test checks
additionally whether information collected by Analyzer phase is stored in
Analysis correctly.

gkossakowski committed:
Do not track names of local and anonymous classes.

For local and anonymous classes we only need to track their products
(corresponding class files). The distinction between local and non local
products has been introduced in ba92ff4ab08f8156c9e6cd676ce81a3ed5702435
but Analyzer phase was not updated to make that distinction in recording
of products.

In order to determine whether a class is local or not, one has to look at
owner chain. This is tricky because Analyzer phase is ran as a late phase
when original owner chains are not available anymore. To overcome that
problem, the LocalToNonLocalClass mapping has been extracted from
Dependency phase and made persistent across the whole compilation run.

The functionality performed by LocalToNonLocalClass resembles what backend
does to produce `EnclosingClass` attributes. Unfortuantely, there's no
stable api we could use to access the information used by Scala compiler's
backend. Hence, we essentially duplicate the functionality.

This commit makes pass both the pending test in ClassNameSpecification
and the scripted recorded-classes test.

gkossakowski committed:
Log only non-empty sets of invalidated files

gkossakowski committed:
Support local classes not coming from src

There are classes seen by Analyzer that are not declared by src. For
example, the implementation class for a trait. We have to gracefully handle
such classes in the Analyzer phase. Classes not seen by the Dependency
phase that populates LocalToNonLocal cache are treated as local.

We make sure that all classes declared in source file have a determined
status by forcing their lookup in the Dependency phase.

gkossakowski committed:
Remove the dead code and unused imports.

Remove some of the the dead code and unused imports in the incremental
compiler's code. This will help with switch from tracking apis at source
level to class level.

gkossakowski committed:
Track API at class level instead of source file level

This commit changes how tracking of API data structures is done within
the incremental compiler. It changes how APIs are passed around and stored
but doesn't change the behavior of the incremental compiler.

Here's a summary of what has changed and what's still being tracked at the
source file level:

  - APIs are tracked per class name in a newly introduced Companions data
    structure; incremental compiler always considers the pair of companions
    from now on
  - only APIs for top level classes are extracted at the moment;
    invalidation is still imprecise
  - Used names are tracked per source file
  - Name hashes are tracked per top-level class (they're part of
    AnalyzedClass data structure)

Companion class and object have to be considered as a pair because they're
given exactly the same name in the incremental compiler. The idea of
naming classes and objects separately has been discussed and rejected
here: https://github.com/sbt/sbt/issues/1104#issuecomment-174192672
APIs of companions are linked together in AnalysisCallback. The ExtractAPI
compiler phase continues to extract apis of classes and objects separately.
More on those changes below.

Most changes in this patch are direct consequences of the changes in the
`interface/other` file. The `Source` has been replaced by the
`AnalyzedClass`. The AnalyzedClass carries an extracted api of the
(class, companion object) pair (stored as `Companions`) plus some meta data
about the pair (e.g. hash sum of its api representation). Source used to
carry both hash sum of the source file contents (hash of the text) and a
hash of the api. The hash of the source file has been introduced to
shortcut equality checking before api hashing has been introduced. Now
it's redundant so it's removed.

The `SourceAPI` used to carry information about packages declared in a
source file but this information wasn't used in the incremental compiler
so its tracking is removed.

I also removed an ugly looking `_internalOnly_` prefixes. The changes in
this branch are not binary or source compatible so it's a good opportunity
to perform some cleanups.

AnalysisCallback has a new method `startSource`. It's needed because we
want to track sources that do not declare any classes and for which there's
no `api` call. The `api` method takes `ClassLike` as an argument and can
be called multiple times per a single source file.

The implementation of the AnalysisCallback has been affected by the switch
from Source to AnalyzedClass tracking the most. The main change here is
that AnalysisCallback is responsible for pairing apis of a companion class
and a companion object. It stores apis of classes and objects separately
and does the same of their name hashes. Right before persisting all that
information it puts both apis into Companions data structure and merges
name hashes for a class and its companion object. Merging is performed
only when both class and object with the same name are declared.
It's worth noting why we cannot just compute name hashes once we have a
class and its companion object available instead of merging precomputed
name hashes. We can't do that because APIs are minimized and only their
hashes are stored so names and signatures of members are lost at this
point. We have to computer name hashes before minimization but then we
don't have both of companions available yet.

For that reason `NameHashing.merge` operation has been introduced that
performs a straightforward merge. NameHashingSpecification provides a basic
test for its properties.

The incremental invalidation algorithm has been affected by switching to
class level api tracking. As a consequence, most of invalidation is
happening at class level now and only right before compilation class names
are mapped to source files that declare them. To emphasize that point,
recompilation of classes has been refactored to its own method
`recompileClasses` in the IncrementalCommon. However, we had two make two
exceptions to invalidation performed on class level:

  1. Sources that just has been added and we don't know their declared
     classes yet so we have to scheduled them for compilation as is.
  2. Sources that do not declare any classes have to be scheduled for
     recompilation directly too

This is the reason why `cycle` takes both invalidated classes and modified
srcs as inputs and why `invalidateInitial` computes both. After the first
iteration of `cycle`, the set of modified sources becomes empty and the
remaining of invalidation is performed at the class level only.

Here's a list of changes I think are worth highlighting either for clarity
or to make a point:

  - SameAPI dropped some old, unused code from TopLevel and NameChanges
    classes
  - APIs do not have any reference to `java.io.File` now, this data
    structure operates purely on class names now
  - helpers methods for looking up dependency information from Relations
    has been changed to work on a class level
  - version number in TextAnalysisFormat has been bumped
  - the `inherited_type_params` scripted test has been removed as it looks
    like not testing anything useful and breaks due to changes to the
    `APIs` interface
  - Analyze doesn't store empty apis for Java source files that do not
    declare any classes; we use `AnalysisCallback.startSource` for tracking
  - The Scala 2.8-specific test has been dropped.
  - The test for Ant-style compilation is marked as pending. Supporting of
    Ant-style compilation is tricky because invalidation is happening at
    the class level now.

gkossakowski committed:
Fix the check for package objects in ExtractAPI

The ExtractAPI would use the wrong check for testing whether a Symbol
corresponds to a package object.

This fixes pending tests for package object invalidation.

gkossakowski committed:
Remove an unnecessary type parameter

Various places of the incremental compiler used to abstract over whether
they operate with source files (for internal dependencies or
class names (for external dependencies). Now that all dependencies are
tracked at class level, we can remove the abstraction.

Also, remove SourceAPIChange that is now unused.

gkossakowski committed:
Small improvement to IncrementalNameHashing

gkossakowski committed:
Remove dead code from API extraction.

gkossakowski committed:
Extract API for each class separately

The ExtractAPI collects API for each class (including inner classes)
separately. For example,

class A {
  class B {
    def foo: Int = 123
  }
}

Is represented as:

// className = A
class A {
  class B
}
// className = A.B
class A.B {
  def foo: Int
}

The signature of an inner classes is represented twice:

  1. as a member of an outer class
  2. as a standalone class

The first representation is needed so dependencies by inheritance are
invalidated properly even if they don't depend explicitly on the inner
class.

The class-based-inheritance test has been expanded to show that changes
to an API of an inner class do not affect classes inheriting from an outer
class.

gkossakowski committed:
Cleanup ExtractUsedNamesSpecification

Remove unused imports and upgrade JUnit runner.

gkossakowski committed:
Use an explicit tree traverser in ExtractUsedNames

Move used names extraction logic to a class that extends Traverser class.
This makes it easier to hold state while traversing the trees.

gkossakowski committed:
Fix compilation erros in JavaCompilerSpec

The compilation erros has been introduced by
800a01460b031790110e2c959ff474d8b9587a1b and, unfortunately, went unnoticed
until now. I cannot rebase because of a non-trivial merge commit that
happened between current commit and 800a01460b031790110e2c959ff474d8b9587a1b.

gkossakowski committed:
Track and extract used names at class level

Switch tracking of used names from source to class level. Changes in this
patch are fairly straightforward and similar to changes to Dependency
extraction phase when switch from source to class level was implemented.
In particular, names mentioned at top level (e.g. in top level imports)
are handled the same as dependencies: they're attributed to the first class
declared in compilation unit.

The class-based-memberRef scripted test that tests a scenario when it
matters if used names are tracked at class level instead of source level.
It is now marked as passing which shows that member ref invalidation with
used names tracked at class level works correctly and optimally.

gkossakowski committed:
Cleanup imports in API phase definition.

gkossakowski committed:
Reduce use of ClassToSourceMapper

Now that we switched to class-based invalidation, mapping between class
names and their defining sources is reduced. E.g. MemberRefInvalidator
doesn't need that capability, it needs just to know whether a given class
name is Scala-defined to decide whether to apply the name hashing
optimization.

This commit introduces a simple predicate and removes use of
the ClassToSourceMapper in context of MemberRefInvalidator.

gkossakowski committed:
Fix bug in SameAPI.sameTopLevel

Change from && to == for checking equality of booleans. It was a typo.

gkossakowski committed:
Persist all extracted apis in APIs datastructure.

We used to persist only top-level apis but now that extraction is done for
all apis, we switch to persisting all of them and remove a TODO from
AnalysisCallback.

gkossakowski committed:
Clarify a binary class name used in ExternalDependency

Clarify that the target class name of an ExternalDependency is a binary
name. The binary instead of source class name is used because the name of
the target class is used for classpath lookups. See
IncrementalCommon.{changedInitial, currentExternalAPI}. The classpath
lookup is performed to get a classpath element corresponding to a binary
class name. After that, another lookup is performed to find a corresponding
Analysis object that tracks given classpath element. I think that with
switch to class-based dependency tracking, a direct lookup for an Analysis
object that tracks a given class name (binary or source) could be done
bypassing a classpath lookup. However, this commit doesn't implement it.

The clarification of source vs binary class name in ExternalDependency
uncovered a bug in `IncrementalCommon.currentExternalAPI` where two types
of class names where confused. The bug is fixed by this commit and
external-src-dependency scripted test is marked as passing.

gkossakowski committed:
Remove unnecessary ExtractAPI.forceStructures calls

We need to call it just once and this is already being done.

gkossakowski committed:
Extract APIs of Java inner classes separately

This is follow up to a3a499535a90df80f65b214d28986da7f2195628 where I
focused on Scala side of the incremental compiler and forgot about APIs
extracted from Java source files. This commit adapts ClassToAPI to extract
APIs of inner classes separately. It expresses APIs of nested classes the
same way as its being done for Scala classes. Each nested class gets a
standalone API object and the enclosing class receives a class with empty
structure as a member.

This commit adapts JavaCompilerForUnitTesting so that extracted APIs of
Java sources can be captured. This functionality is used in a newly
introduced ClassToAPISpecification. The test dependency from api to
classfile subproject is introduced so JavaCompilerForUnitTesting is
visible in the api subproject's tests.

gkossakowski committed:
Fix computing name hashes for private classes

Documentation TBD.

gkossakowski committed:
Fix handling of names of Java-defined class names

Documentation to be done. TODO: handle local classes properly.

gkossakowski committed:
Improve logging of invalidation of package objects

Do not log the transitive walk performed to find package objects that
inherit (possibly indirectly) from invalidated classes. Log just the
package objects found thanks to transitive walk. In most scenarios, no
package objects are involved so nothing is being logged.

gkossakowski committed:
Merge branch '1.0' into cd-filter-branch

This commit merges 1.0 branch into class-based-dependencies branch.

The class-based-dependencies branch has been obtained from the sbt repo by
rerunning the filter-branch command that has been used originally to
spawn zinc repo out of sbt repo:

git filter-branch --index-filter 'git rm --cached -qr -- . && git reset -q $GIT_COMMIT -- compile interface util/classfile util/classpath util/datatype util/relation' --prune-empty

That preserved some merges from the original class-based-dependencies
branch that didn't make sense so I rebased the filtered branch on top of
7e9ccc10af0c53f17ecfe87cead3ef03fd95642f. This commit is the first commit
in zinc's history after filtering. Once the rebase was done and I got a
cleaner history, I've merged 1.0 into my class-based-dependencies branch.
This commit is exactly that merge.

The merge itself is very complicated because very non-trivial changes
happened in both zinc and my class-based-dependencies branch. What you see
here represents the best effort at merging. I expect mistakes to be
discovered and fixed in subsequent commits. I got it as far as to compile
everything and make most of the test pass.

I highlight below some of the least obvious changes I've made:

  - Analysis.groupBy operation has been nuked. It's been broken in
    class-based-dependencies and it's not clear who needs it
  - AnalysisTest is removed as it was mostly testing the Analysis.groupBy
  - changes to definition.json and other.json are ports of changes from
    definition and other (from the interface subproject). Consult hisotry
    of those files to understand those changes
  - TestAnalyzingCompiler has been adapted to handle invalidation at class
    level instead of source file level
  - TestAnalysisCallback has been adapted to changes how apis, products
    and dependencies are tracked. A lot of code has been copied over from
    AnalysisCallback. I'm not sure if that the best course of action but
    I don't understand the role of TestAnalysisCallback exactly to have
    a better opinion.


# Conflicts:
#	compile/api/src/main/scala/xsbt/api/ShowAPI.scala
#	compile/api/src/test/scala/xsbt/api/NameHashingSpecification.scala
#	compile/inc/src/main/scala/sbt/inc/Dependency.scala
#	compile/integration/src/main/scala/sbt/compiler/AggressiveCompile.scala
#	compile/integration/src/main/scala/sbt/compiler/IncrementalCompiler.scala
#	compile/interface/src/main/scala/xsbt/Dependency.scala
#	compile/persist/src/main/scala/sbt/inc/AnalysisFormats.scala
#	compile/persist/src/main/scala/xsbt/api/AnalyzedClassFormat.scala
#	compile/persist/src/main/scala/xsbt/api/SourceFormat.scala
#	compile/src/test/scala/sbt/compiler/javac/JavaCompilerSpec.scala
#	incrementalcompiler/src/main/scala/sbt/internal/inc/MixedAnalyzingCompiler.scala
#	incrementalcompiler/src/main/scala/sbt/internal/inc/javac/AnalyzingJavaCompiler.scala
#	interface/definition
#	interface/other
#	interface/src/main/java/xsbti/DependencyContext.java
#	interface/src/test/scala/xsbti/TestCallback.scala
#	internal/compiler-bridge/src-2.10/main/scala/xsbt/API.scala
#	internal/compiler-bridge/src-2.10/main/scala/xsbt/CompilerInterface.scala
#	internal/compiler-bridge/src-2.10/main/scala/xsbt/ExtractAPI.scala
#	internal/compiler-bridge/src-2.10/main/scala/xsbt/ExtractUsedNames.scala
#	internal/compiler-bridge/src/test/scala/xsbt/DependencySpecification.scala
#	internal/compiler-bridge/src/test/scala/xsbt/ExtractAPISpecification.scala
#	internal/compiler-bridge/src/test/scala/xsbt/ExtractUsedNamesSpecification.scala
#	internal/compiler-bridge/src/test/scala/xsbt/ScalaCompilerForUnitTesting.scala
#	internal/compiler-interface/src/main/java/xsbti/AnalysisCallback.java
#	internal/incrementalcompiler-apiinfo/src/main/scala/xsbt/api/APIUtil.scala
#	internal/incrementalcompiler-apiinfo/src/main/scala/xsbt/api/HashAPI.scala
#	internal/incrementalcompiler-apiinfo/src/main/scala/xsbt/api/NameHashing.scala
#	internal/incrementalcompiler-classfile/src/main/scala/sbt/internal/inc/classfile/Analyze.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/APIDiff.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/APIs.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/Analysis.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/Compile.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/Incremental.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/IncrementalAntStyle.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/IncrementalDefaultImpl.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/IncrementalNameHashing.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/MemberRefInvalidator.scala
#	internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/Relations.scala
#	internal/incrementalcompiler-core/src/test/scala/sbt/inc/AnalysisTest.scala
#	internal/incrementalcompiler-core/src/test/scala/sbt/inc/TestCaseGenerators.scala
#	internal/incrementalcompiler-persist/src/main/scala/sbt/internal/inc/TextAnalysisFormat.scala
#	internal/incrementalcompiler-persist/src/main/scala/xsbt/api/SourceFormat.scala
#	internal/incrementalcompiler-persist/src/test/scala/sbt/inc/TextAnalysisFormatSpecification.scala

gkossakowski committed:
Add .gitignore file

gkossakowski committed:
Fix broken merge in Dependency.scala

The merge in b0a8a91aaa5b6f55714986f7bf389ec8aa091ec0 messed up code
in Dependency.scala. This commit just copies over the code of the
`traverse` method as-is from class-based-dependencies.

This commit makes all tests in compilerBridge to pass.

gkossakowski committed:
Fix build failure of the Scala 2.10 bridge sources

The fix was to just copy over the main code and adapt it to Scala 2.10
compiler API.

gkossakowski committed:
Fix Scaladoc failures in Dependency.scala

gkossakowski committed:
Zinc is compatible with Scala 2.10 and 2.11 only

The class-based-dependencies branch nuked compatibility with older Scala
compiler versions.

gkossakowski committed:
Revert "Include private members in API hash of traits."

This reverts commit 7956abafdc6da7f4d639db1702b8b216bf20f6bc.

And mark scripted tests that are affected by the revert as pending.

This is a work-around for the sbt/sbt#2490. The revert got lost in my
rebase of the original class-based-dependencies branch.

# Conflicts:
#	internal/incrementalcompiler-apiinfo/src/main/scala/xsbt/api/HashAPI.scala
#	internal/incrementalcompiler-apiinfo/src/test/scala/xsbt/api/NameHashingSpecification.scala

gkossakowski committed:
Fix import-class scripted test

By porting changes from
gkossakowski/sbt@d91aa837af018f44ece6a95a051279d37c414c38

gkossakowski committed:
Merge remote-tracking branch 'sbt/1.0' into class-dependencies

gkossakowski committed:
Remove old incremental compilation algorithm implementation.

We do not support it in a class-based dependency tracking.
Remove tests that are specific to the old incremental compiler and are
failing now.

This is a port of gkossakowski/sbt@5c5fb6dfaa567ce8f1e6af80bba126e7e50ea941

gkossakowski committed:
Make scripted commands work with classes instead of srcs

Adapt checkDependencies and checkRecompilations to work with classes
instead of source files.

Mark check-products scripted test as pending. It looks like it's failing
for a legitimate reason that will be investigated in a separate commit.

gkossakowski committed:
Fix specify-inc-options scripted test

Turning off name hashing is not supported anymore.

gkossakowski committed:
Add documentation to DependencyContext

Per discussion in #81, I know how to do it now.

gkossakowski committed:
Port class-based-inheritance scripted test

From gkossakowski/sbt@c08e430fdf8bfd95d5848563937eb6c033072aa6

gkossakowski committed:
Fix checkProducts and introduce checkClasses

The previous version of checkProducts was in fact checking classes that
are declared in a source file thus this command has been renamed to
checkClasses.

The proper implementation of checkProducts has been introduced too.

gkossakowski committed:
Port recorded-products scripted test.

From gkossakowski/sbt@4e513ca7918f88f4a0481ec9629b40d33defd37b

gkossakowski committed:
Port java-inner scripted test

From gkossakowski/sbt@ba92ff4ab08f8156c9e6cd676ce81a3ed5702435
This test is marked as pending due to #83

gkossakowski committed:
Port local-class-inheritance scripted test

From gkossakowski/sbt@ce0d24ad04f5ceb58a41deb4c0b1462b174d3ad1

gkossakowski committed:
Port gkossakowski/sbt@800a01460b031790110e2c959ff474d8b9587a1b

Not all changes are ported because scripted doesn't run inherited-macros
test.

gkossakowski committed:
Add .DS_Store to ignored files

gkossakowski committed:
Port class-based-memberRef scripted test

From gkossakowski/sbt@5737d6ef2bef6d4510b99cd7dea87000bac42494

gkossakowski committed:
Port gkossakowski/sbt@e09fa6d383f2fd9017954af6805b5e129f64047e

Tests mark as pending due to #83.

gkossakowski committed:
Port external-src-dependency sripted test

Marked as pending due to #84.

gkossakowski committed:
Mark java scripted tests as passing

Those test do not test anything due to #83 but they do not fail either so
they cannot be marked as pending.

gkossakowski commented:
Review by @Duhemm and @adriaanm 

Realistically speaking, I don't expect you to review all commits. I'd recommend you going through the list of tasks that I implemented and just check if you understand why they were done. If something is not clear, ask me.

If the motivation coming from tasks is clear then commits will be easy to understand both now and in the future.

smarter commented on internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/MemberRefInvalidator.scala:
```diff
@@ -53,65 +53,59 @@ import xsbt.api.APIUtil
  * of regular members then we'll invalidate sources that use those names.
  */
```
The whole comment for this class needs to be updated since it describes the strategy as being implemented as `T => Set[File]` whereas now it's `String => Set[String]`.

gkossakowski replied:
I've pushed a commit that updates docs.

Duhemm commented on internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/Relations.scala:
```diff
@@ -196,35 +265,38 @@ trait Relations {
+      FSRelationDescriptor("direct external dependencies", r => Relations.emptySource.external),
+      // NameHashing doesn't provide public inherited dependencies
+      FFRelationDescriptor("public inherited source dependencies", r => Relations.emptySource.internal),
+      // NameHashing doesn't provide public inherited dependencies
```
If NameHashing is becoming the only option, we should be able to get rid of `direct` and `publicInherited` relations

gkossakowski replied:
Good idea. I've removed them and while I was at it, I removed `declaredClasses` relations too (that wasn't used for anything).

Duhemm commented on internal/compiler-bridge/src-2.10/main/scala/xsbt/Dependency.scala:
```diff
@@ -33,59 +33,150 @@ final class Dependency(val global: CallbackGlobal) extends LocateClassFile {
 
   def newPhase(prev: Phase): Phase = new DependencyPhase(prev)
-  private class DependencyPhase(prev: Phase) extends GlobalPhase(prev) {
+  private class DependencyPhase(prev: Phase) extends Phase(prev) {
```
Isn't this undoing the changes introduced in sbt/sbt#2435 / #73 ?

Duhemm commented on internal/incrementalcompiler-core/src/main/scala/sbt/internal/inc/IncrementalCommon.scala:
```diff
@@ -106,21 +137,22 @@ private[inc] abstract class IncrementalCommon(log: sbt.util.Logger, options: Inc
     }
-  def sameSource[T](src: T, a: Source, b: Source): Option[APIChange[T]] = {
+  def sameClass(className: String, a: AnalyzedClass, b: AnalyzedClass): Option[APIChange] = {
     // Clients of a modified source file (ie, one that doesn't satisfy `shortcutSameSource`) containing macros must be recompiled.
```
The comment should read `shortcutSameClass` now that it has been renamed

gkossakowski replied:
Fixed.

Duhemm commented:
It looks like `ClassToAPISpecification.scala` has been deleted in b0a8a91aaa5b6f55714986f7bf389ec8aa091ec0

What this intended?

gkossakowski committed:
Extend local class inheritance scripted test

Extend the local class inheritance test to check for two properties of
invalidation by local inheritance:

  1. Local inheritance are invalidated in the same step (iteration) as
     regular inheritance dependencies
  2. Local inheritance invalidation is not transitive

The current behavior is suboptimal because 1. doesn't hold as shown by
the extended test case.

gkossakowski committed:
Improve invalidation of local inheritance dependencies.

There two properties of local inheritance invalidation:

1. Local inheritance are invalidated in the same step (iteration) as
regular inheritance dependencies
2. Classes invalidated by local inheritance are terminal: they are not
   considered for the transitive inheritance invalidation or the member
   reference invalidation

Before this commit the first property didn't hold. We're fixing it by
extending invalidation by local inheritance to consider all classes
invalidated by regular inheritance.

gkossakowski committed:
Persist the local inheritance relation

The 47416b394e3730d67367da9f011f8ef3de23b576 introduced the local
inheritance relation but forgot about adding it persistence code.
This commit fixes that mistake and extends TestCaseGenerators to generate
local inheritance relations and thus test Analysis persistence correctly.

gkossakowski committed:
Move Analyze and ClassToAPI specifications

The b0a8a91aaa5b6f55714986f7bf389ec8aa091ec0 merge commit didn't resolve
merge conflicts for AnalyzeSpecification.scala and for
ClassToAPISpecifaction.scala. This commit oves them to the right
subprojects and adapts them to Scalatest APIs.

Both specifications depend on TestCallback so it had to be moved to a
subproject commonly referenced. I've moved it to the compiler-interface
subproject. As a consequence, I had to add a dependency on scala-library
in `test` configuration to compile TestCallback written in Scala.

gkossakowski commented:
> It looks like ClassToAPISpecification.scala has been deleted in b0a8a91

It's a broken conflict resolution that I've missed. Great catch! I pushed a commit that moves ClassToAPI and Analyze specifications.

I also pushed commits that fix local inheritance dependencies.

gkossakowski committed:
Update docs of Incremental{Common, NameHashing}

Update documentation and logging statement to talk about classes instead
of source files.

gkossakowski committed:
Remove `declaredClasses` relation.

The `declaredClasses` relation duplicates the data stored and functionality
provided by `classes` relation and it's not being used in incremental
compiler. Let's remove it.

A historical note: the declaredClasses relation has been introduced at
the time when classes relation fulfilled a different role. At some point,
classes has been refactored to fulfill exactly the same role as
declaredClasses relation.

gkossakowski committed:
Remove `direct` and `publicInherited` relations

These relations supported the incremental compilation when name hashing
was turned off. Since turning off name hashing is not supported anymore
we can remove those relations and code supporting them.

Also, Analysis.merge has been removed. The code in `merge` method never
supported turning off name hashing and now is a dead code.

gkossakowski committed:
Update docs of MemberRefInvalidator

Update documentation of MemberRefInvalidator to talk about invalidation
of classes instead of source files. Also, I removed explanation of type
parametricity of the invalidator that has been removed already.

gkossakowski commented:
Thanks for spotting all the missing articles @smarter. I pushed updated version of that commit.

smarter commented:
The comment on top of `ExtractAPI.scala` is no longer accurate, it says:

```
 * This class depends on instance of CallbackGlobal instead of regular Global because
 * it has a call to `addInheritedDependencies` method defined in CallbackGlobal.
```

But the call to `addInheritedDependencies` is gone.

gkossakowski committed:
Refactor classpath and analysis lookups.

Introduce the `Lookup` trait that contains all necessary methods that are
needed by zinc to perform classpath and Analysis lookups for external
dependencies.

Adapt existing logic to make use of the newly introduced interface.
We have several different variants of `lookupAnalysis` because different
parts of zinc start either with just a binary class name or a class file
or both. Eventually we should perform all lookups based just on a binary
class name.

The related sbt/sbt#2525 ticket describes how looking up Analysis could
be sped up and improve `compile` performance significantly.

gkossakowski committed:
Reduce dependency on CallbackGlobal.

Remove dependency on CallbackGlobal from ExtractAPI that doesn't require
it anymore and update the docs.

Remove dependency on CallbackGlobal from ClassName and GlobalHelpers that
do not need it either.

gkossakowski commented:
I added a small refactoring of the code that performs classpath and Analysis lookups. My motivation was streamline bits of code that are passing `entry` and `forEntry` functions.

I updated `ExtractAPI` per @smarter's comment.

Is this good to go for you?

smarter commented:
LGTM in the limited testing I did, however I'm still a bit concerned by the performance regressions on scala/scala (https://github.com/sbt/sbt/issues/1104#issuecomment-199360458), would be good to know why this happens and if this happens in other big projects too.

smarter commented:
Also the tests seem to be failing, no idea why

Duhemm commented:
LGTM!

gkossakowski committed:
Merge remote-tracking branch 'sbt/1.0' into class-dependencies

# Conflicts:
#	.gitignore

gkossakowski commented:
The test failure was a Travis glitch.

eed3si9n commented:
Merged!

gkossakowski commented:
Excellent news! I'll work on remaining issues. Just as a convenience, here they are:
- handling of anonymous and local classes defined in Java (similar logic will have to be implemented as for handling local Scala classes)
- Java-specific tests are ignored (see #83)
- asSeenFrom is not honored correctly for inherited classes (they are extracted only as seen from declaration site) which leads to under compilation and missing typechecking errors (https://gitter.im/scala/contributors?at=56ec053bbb4a1731739afb6f)

retronym commented on internal/compiler-bridge/src-2.10/main/scala/xsbt/ExtractUsedNames.scala:
```diff
@@ -57,57 +84,81 @@ class ExtractUsedNames[GlobalType <: CallbackGlobal](val global: GlobalType) ext
+      // to types but that might be a bad thing because it might expand aliases eagerly which
+      // not what we need
+      case t: TypeTree if t.original != null =>
+        t.original.foreach(traverse)
```
I believe this line (and the one above for macro originals) is a performance regression. Both `foreach` and `traverse` will recurse into the tree. See https://github.com/sbt/zinc/pull/187

Either `t.original.foreach(handleTreeNode)` or simply `traverse(t.original)` seems to be the right approach.

jvican replied:
FTR, fixed in https://github.com/sbt/zinc/pull/193.
